### ISSUE TITLE
Inaccurate code coverage output

### ISSUE THREAD

**Issue by rolaveric** [guybedford/karma-systemjs-experiment#1]
Testing with https://github.com/rolaveric/ng1-to-ng2-example/tree/karma-systemjs-experiment
I'm getting a strange issue where it reports that the browser has disconnected, but after it reports back that the tests executed successfully.

Karma output:

```
29 04 2016 21:53:31.741:INFO [launcher]: Starting browser Chrome
29 04 2016 21:53:35.562:INFO [Chrome 50.0.2661 (Windows 7 0.0.0)]: Connected on socket /#d7Hcav8NSJveZBU9AAAA with id 54122703
Chrome 50.0.2661 (Windows 7 0.0.0): Executed 2 of 2 SUCCESS (0 secs / 0 secs)
Missing error handler on `socket`.
Error: Unable to lookup source for C:\Users\rolaveric\WebstormProjects\ng1-to-ng2-example\app\model\login.js
    at remapIstanbul.readFile (C:\Users\rolaveric\WebstormProjects\ng1-to-ng2-example\node_modules\karma-systemjs\lib\reporter.js:23:17)
    at C:\Users\rolaveric\WebstormProjects\ng1-to-ng2-example\node_modules\remap-istanbul\lib\remap.js:189:39
    at Array.forEach (native)
    at C:\Users\rolaveric\WebstormProjects\ng1-to-ng2-example\node_modules\remap-istanbul\lib\remap.js:182:22
    at Array.forEach (native)
    at remap (C:\Users\rolaveric\WebstormProjects\ng1-to-ng2-example\node_modules\remap-istanbul\lib\remap.js:181:12)
    at remap (C:\Users\rolaveric\WebstormProjects\ng1-to-ng2-example\node_modules\karma-systemjs\lib\reporter.js:16:21)
    at onBrowserComplete (C:\Users\rolaveric\WebstormProjects\ng1-to-ng2-example\node_modules\karma-systemjs\lib\reporter.js:6:25)
    at null.<anonymous> (C:\Users\rolaveric\WebstormProjects\ng1-to-ng2-example\node_modules\karma\lib\events.js:13:22)
    at emitTwo (events.js:105:20)
    at emit (events.js:185:7)
    at onComplete (C:\Users\rolaveric\WebstormProjects\ng1-to-ng2-example\node_modules\karma\lib\browser.js:142:13)
    at Socket.<anonymous> (C:\Users\rolaveric\WebstormProjects\ng1-to-ng2-example\node_modules\karma\lib\events.js:13:22)
    at emitTwo (events.js:105:20)
    at Socket.emit (events.js:185:7)
    at Socket.onevent (C:\Users\rolaveric\WebstormProjects\ng1-to-ng2-example\node_modules\socket.io\lib\socket.js:335:8)
    at Socket.onpacket (C:\Users\rolaveric\WebstormProjects\ng1-to-ng2-example\node_modules\socket.io\lib\socket.js:295:12)
    at Client.ondecoded (C:\Users\rolaveric\WebstormProjects\ng1-to-ng2-example\node_modules\socket.io\lib\client.js:193:14)
    at Decoder.Emitter.emit (C:\Users\rolaveric\WebstormProjects\ng1-to-ng2-example\node_modules\component-emitter\index.js:134:20)
    at Decoder.add (C:\Users\rolaveric\WebstormProjects\ng1-to-ng2-example\node_modules\socket.io-parser\index.js:247:12)
    at Client.ondata (C:\Users\rolaveric\WebstormProjects\ng1-to-ng2-example\node_modules\socket.io\lib\client.js:175:18)
    at emitOne (events.js:90:13)
29 04 2016 21:53:57.351:WARN [Chrome 50.0.2661 (Windows 7 0.0.0)]: Disconnected (1 times)
Chrome 50.0.2661 (Windows 7 0.0.0): Executed 2 of 2 DISCONNECTED (10.021 secs / 0 secs)
```

The browser itself doesn't think it's disconnected (Or at least shows no error messages or console output), but it's definitely not responding to changes anymore.

Karma output after a change:

```
29 04 2016 21:59:58.168:INFO [watcher]: Changed file "C:/Users/rolaveric/WebstormProjects/ng1-to-ng2-example/app/model/login.js".
29 04 2016 21:59:58.449:WARN [karma]: No captured browser, open http://localhost:9876/
```

**Comment by rolaveric** [comment]
It's definitely related to the `systemjs-remap-coverage` reporter.
When I disable it in the Karma config, the issue goes away.

If you can't reproduce it, it could be a Windows specific issue (I really need to setup my Ubuntu dual-boot again...)

**Comment by guybedford** [comment]
@rolaveric yes this definitely looks like a windows issue.

**Comment by guybedford** [comment]
It tries to find the source file in https://github.com/guybedford/karma-systemjs-experiment/blob/master/lib/reporter.js#L19, which may not be matching up due to windows pathing differences. Perhaps just replacing `/` with `\` in that would work.

**Comment by rolaveric** [comment]
Yep. I see it. `servedFiles` already has  the slashes changed to `/` but `filePath` is using `\`.
I'll test out a fix and create a PR.

**Comment by guybedford** [comment]
👍

**Comment by rolaveric** [comment]
Created PR #3 
But I noticed that if I add code to `app/model/login.js`, code which isn't being tested, the code coverage report doesn't reflect that.
The source is the same, but it's saying 100% coverage.
Are you getting the same or can we chalk this down to Windows again?

**Comment by guybedford** [comment]
Would be interesting to see the coverage data here. The reverse-source-maps process isn't completely accurate, but gets the 80% from what I've seen.

**Comment by rolaveric** [comment]
Alright, I'll dig into it tomorrow.

Not sure if this is what you're looking for but this is the JSON coverage report:

```
{
  "app\\model\\login.js": {
    "path": "app\\model\\login.js",
    "statementMap": {
      "1": {
        "start": {
          "line": 1,
          "column": 0
        },
        "end": {
          "line": null,
          "column": -1
        }
      },
      "2": {
        "start": {
          "line": 2,
          "column": 0
        },
        "end": {
          "line": null,
          "column": -1
        }
      }
    },
    "fnMap": {
      "1": {
        "name": "(anonymous_7)",
        "line": 5,
        "loc": {
          "start": {
            "line": 5,
            "column": 13
          },
          "end": {
            "line": null,
            "column": -1
          }
        }
      },
      "2": {
        "name": "(anonymous_11)",
        "line": 23,
        "loc": {
          "start": {
            "line": 23,
            "column": 17
          },
          "end": {
            "line": null,
            "column": -1
          }
        }
      }
    },
    "branchMap": {},
    "s": {
      "1": 2,
      "2": 1
    },
    "b": {},
    "f": {
      "1": 1,
      "2": 2
    },
    "code": [
      "import {Class} from 'angular2/core';\r\nimport {Http} from 'angular2/http';\r\n\r\n// Model for handling login\r\nexport class LoginModel {\r\n  static get parameters() {\r\n    return [[Http]];\r\n  }\r\n\r\n  constructor(http) {\r\n    this.http = http;\r\n    this.username = null;\r\n  }\r\n\r\n  /**\r\n   * Handles login.\r\n   * @param username {string}\r\n   * @param password {string}\r\n   * @returns {Promise}\r\n   */\r\n  login(username, password) {\r\n    return this.http.post('/api/v1/login', {username: username, password: password})\r\n      .subscribe(() => this.username = username);\r\n  }\r\n\r\n  /**\r\n   * Deliberately unused for testing code coverage.\r\n   */\r\n  unusedMethod() {\r\n    console.log('meep');\r\n  }\r\n}\r\n\r\nexport const LoginModelProvider = Class({constructor: LoginModel});\r\n"
    ],
    "l": {
      "1": 2,
      "2": 1
    }
  }
}
```

**Comment by rolaveric** [comment]
Ok, finally got some time to look at all of this.

I just did a fresh clone of [ng1-to-ng2-example/karma-systemjs-experiment](https://github.com/rolaveric/ng1-to-ng2-example/tree/karma-systemjs-experiment) on Ubuntu with NodeJS v6.1 and I'm getting the exact same output I do on Windows.
The import statements have code coverage, but nothing else does. eg. In the screenshot below the `unusedMethod()` should show up in red as not covered.

![image](https://cloud.githubusercontent.com/assets/960670/15274725/45adae2c-1afd-11e6-848c-bcd406f29e24.png)

Once this is resolved I think we can start looking at merging changes into `karma-systemjs`.
I can start by going through the existing issues in `karma-systemjs` to see if they need to be applied to the code, have test scenarios created (eg. Different transpilers), or just make sure they're documented in the README.

**Comment by guybedford** [comment]
Thanks so much @rolaveric for taking another look at this. I think this source maps issue has been resolved in the latest jspm 0.17 and plugin-babel releases as there was a source maps issue there and the coverage report has improved in my latest tests.

I have since encapsulated the Istanbul support into its own project at https://github.com/guybedford/systemjs-istanbul.

This would allow the preprocessor to be written identically using `builder.compile` as in https://github.com/guybedford/karma-systemjs-experiment/blob/master/lib/preprocessor.js#L19, but instead have the coverage handled by injecting into SystemJS builder with:

``` javascript
require('systemjs-istanbul').hookSystemJS(builder.loader);
```

The coverage object can then be pased via `require('systemjs-istanbul').remapCoverage(COVERAGE_OBJECT)` to get the source-mapped coverage file.
